src/plugins/discover/public/application/components/chart/download_csv_button/download_csv_button.tsx

```diff
@@ -0,0 +1,131 @@
+import moment from 'moment';
+import { OpenSearchSearchHit } from '../../../doc_views/doc_views_types';
+import { IndexPattern, UI_SETTINGS } from '../../../../../../data/common';
+import { getServices } from '../../../../opensearch_dashboards_services';
+import { DiscoverRootState, useSelector } from '../../../utils/state_management';
+import { buildColumns } from '../../../utils/columns';
+import { DOC_HIDE_TIME_COLUMN_SETTING } from '../../../../../common';
+import { getLegacyDisplayedColumns } from '../../default_discover_table/helper';
+
+export interface DownloadCsvButtonProps {
+  indexPattern: IndexPattern;
+  rows: OpenSearchSearchHit[];
+}
+
+export const selectColumnNames = ({
+  indexPattern,
+  hideTimeColumn,
+  isShortDots,
+}: {
+  indexPattern: IndexPattern;
+  hideTimeColumn: boolean;
+  isShortDots: boolean;
+}) => (state: DiscoverRootState) => {
+  const stateColumns = state.discover.columns;
+  // check if state columns is not undefined, otherwise use buildColumns
+  const columns = buildColumns(stateColumns || []);
+
+  // Handle the case where all fields/columns are removed except the time-field one
+  const adjustedColumns =
+    columns.length === 1 && columns[0] === indexPattern.timeFieldName
+      ? [...columns, '_source']
+      : columns;
+
+  const displayedColumns = getLegacyDisplayedColumns(
+    adjustedColumns,
+    indexPattern,
+    hideTimeColumn,
+    isShortDots
+  );
+  return displayedColumns.map((column) => column.name);
+};
+
+const formatRowsForCsv = ({
+  displayedColumnNames,
+  indexPattern,
+}: {
+  displayedColumnNames: string[];
+  indexPattern: IndexPattern;
+}) => (row: OpenSearchSearchHit) => {
+  return displayedColumnNames.map((colName) => {
+    const fieldInfo = indexPattern.fields.getByName(colName);
+
+    if (typeof row === 'undefined') {
+      return '';
+    }
+
+    if (fieldInfo?.type === '_source') {
+      const formattedRow = indexPattern.formatHit(row, 'text');
+      return JSON.stringify(formattedRow);
+    }
+
+    const formattedValue = indexPattern.formatField(row, colName, 'text');
+
+    if (typeof formattedValue === 'undefined') {
+      return '';
+    }
+
+    return formattedValue;
+  });
+};
+
+export const DownloadCsvButton = ({ indexPattern, rows }: DownloadCsvButtonProps) => {
+  const { uiSettings } = getServices();
+
+  const [isShortDots, hideTimeColumn] = useMemo(() => {
+    return [
+      uiSettings.get(UI_SETTINGS.SHORT_DOTS_ENABLE),
+      uiSettings.get(DOC_HIDE_TIME_COLUMN_SETTING),
+    ];
+  }, [uiSettings]);
+
+  const displayedColumnNames = useSelector(
+    selectColumnNames({ indexPattern, hideTimeColumn, isShortDots })
+  );
+
+  const handleDownloadCsv = (): void => {
+    const csvRowData = rows.map(formatRowsForCsv({ displayedColumnNames, indexPattern }));
+
+    const csvData = Papa.unparse(
+      {
+        fields: displayedColumnNames,
+        data: csvRowData,
+      },
+      {
+        quotes: uiSettings.get('csv:quoteValues', true),
```

Comment:
for consistency I think we should had a constant and access it like `DOC_HIDE_TIME_COLUMN_SETTING` for these two

Reply:
do you mean the `csv:quoteValues` string?